sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -129,13 +159,39 @@ case class GpuCaseWhen(
     }
   }
 
-  override def columnarEval(batch: ColumnarBatch): Any = {
-    // `elseRet` will be closed in `computeIfElse`.
-    val elseRet = elseValue
-      .map(_.columnarEval(batch))
-      .getOrElse(GpuScalar(null, branches.last._2.dataType))
-    branches.foldRight[Any](elseRet) { case ((predicateExpr, trueExpr), falseRet) =>
-      computeIfElse(batch, predicateExpr, trueExpr, falseRet)
+  @transient
+  private[this] lazy val trueExpressions = branches.map(_._2)
+
+  override def columnarEval(batch:ColumnarBatch): Any = {
+    val size = branches.size
+    val predictions = new Array[GpuColumnVector](size)
+    var isAllPredsFalse = true
+    var i = 0
+
+    withResource(predictions) { preds =>
+      while (i < size) {
+        // If any predication is the first all-true, then evaluate its true expression
+        // and return the result.
+        preds(i) = GpuExpressionsUtils.columnarEvalToColumn(branches(i)._1, batch)
+        val p = preds(i)
+        if (isAllPredsFalse && isAllTrue(p)) {
+          return trueExpressions(i).columnarEval(batch)
+        }
+        isAllPredsFalse = isAllPredsFalse && isAllFalse(p)
+        i += 1
+      }
+
+      val elseRet = elseValue
+        .map(_.columnarEval(batch))
+        .getOrElse(GpuScalar(null, branches.last._2.dataType))
+      if (isAllPredsFalse) {
+        // No predication has a true, so return the else value.
+        elseRet
```

Comment:
We can't return a raw `GpuScalar` here, this isn't a reduction expression.  It needs to be a `GpuColumnVector` since the input being evaluated for the entire expression was a `ColumnarBatch`.  That probably explains the failing test case in the CI run.  Before this code change, that was automatically handled by `ColumnView.ifElse` which supports taking a scalar directly.

I think this needs to be something like:
```suggestion
        elseRet match {
          case s: GpuScalar =>
            withResource(s) { s =>
              GpuColumnVector.from(s, batch.numRows, s.dataType)
            }
          case c => c
        }
```

Reply:
Great catch, updated. also updated this in `GpuIf`.